### PR TITLE
Several fixes

### PR DESCRIPTION
**Pull request by FreyJo**
- StepEquilibrationMode.DIRECT fix broke test unfortunately.
- move w0 to problem
- add options DIRECT_COMPLEMENTARITY, ALL_XCURRENT_WOPT_PREV
- make enums auto()

### CODE REVIEW

File: nosnoc/nosnoc.py [hunk 1]
```diff
@@ -1056,17 +1062,20 @@ def __init__(self, opts: NosnocOpts, model: NosnocModel, ocp: Optional[NosnocOcp
             print(f"{opts=}")
             print("\nerror creating solver for problem above:\n")
             print(f"\nerror is \n\n: {err}")
-            import pdb
-            pdb.set_trace()
+            breakpoint()
```
Comment (apozharski) [inline nosnoc/nosnoc.py:1065]: do we really want to drop into PDB here?

File: nosnoc/nosnoc.py [hunk 2]
```diff
@@ -1056,17 +1062,20 @@ def __init__(self, opts: NosnocOpts, model: NosnocModel, ocp: Optional[NosnocOcp
...
 
     def initialize(self):
         opts = self.opts
         prob = self.problem
         x0 = prob.model.x0
 
-        if opts.initialization_strategy == InitializationStrategy.ALL_XCURRENT_W0_START:
+        # print("\nw \t\t\t w0 \t\t w0_original:")
```
Comment (apozharski) [inline nosnoc/nosnoc.py:1072]: prints :)

File: nosnoc/nosnoc.py [hunk 3]
```diff
@@ -1201,6 +1209,21 @@ def solve(self) -> dict:
...
+        if opts.print_level > 1 and opts.constraint_handling == ConstraintHandling.LEAST_SQUARES:
+            threshold = np.max([np.sqrt(cost_val) / 10, opts.comp_tol])
+            g_val = prob.g_fun(w_opt, p_val).full().flatten()
+            if max(abs(g_val)) > threshold:
+                print("\nconstraint violations:")
+                for ii in range(len(g_val)):
+                    if g_val[ii] > threshold:
+                        print(f"g_val[{ii}] = {g_val[ii]} expr: {prob.g_lsq[ii]}")
+
+        # if cost_val > opts.comp_tol * 1e2:
```
Comment (apozharski) [inline nosnoc/nosnoc.py:1222]: This can probably go as well.

File: nosnoc/nosnoc.py [hunk 4]
```diff
@@ -945,8 +951,12 @@ def __init__(self, opts: NosnocOpts, model: NosnocModel, ocp: Optional[NosnocOcp
         self.comp_res = Function('comp_res', [self.w, self.p], [J_comp])
         self.g_fun = Function('g_fun', [self.w, self.p], [self.g])
 
+        # copy original w0
+        self.w0_original = copy(self.w0)
```
Comment (apozharski) [inline nosnoc/nosnoc.py:955]: Is this used anywhere?

File: nosnoc/nosnoc.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 from typing import Optional, List
 from abc import ABC, abstractmethod
 import time
+from copy import copy
```
Comment (apozharski) [inline nosnoc/nosnoc.py:4]: If the copy is not necessary this can probably also go.

**Review by apozharski — CHANGES_REQUESTED** [review]
A couple of small things.

File: nosnoc/nosnoc.py (hunk 2 above)
Comment (FreyJo) [reply nosnoc/nosnoc.py:1072]: removed

File: nosnoc/nosnoc.py (hunk 4 above)
Comment (FreyJo) [reply nosnoc/nosnoc.py:955]: It is rather useful to keep it to reset dual variables to something neutral

File: nosnoc/nosnoc.py (hunk 3 above)
Comment (FreyJo) [reply nosnoc/nosnoc.py:1222]: done

File: nosnoc/nosnoc.py (hunk 4 above)
Comment (FreyJo) [reply nosnoc/nosnoc.py:955]: Indeed the functionality is missing, but I'd like to keep it

File: nosnoc/nosnoc.py (hunk 1 above)
Comment (FreyJo) [reply nosnoc/nosnoc.py:1065]: We also did before. I just change to breakpoint. We can also just raise the exception of course.

Comment (apozharski) [reply nosnoc/nosnoc.py:1065]: I would raise an exception but that can be changed later.

File: nosnoc/nosnoc.py (hunk 4 above)
Comment (apozharski) [reply nosnoc/nosnoc.py:955]: Got it.

**Review by apozharski — APPROVED** [review]
LGTM

**Pull request merged**